### PR TITLE
[ADD] OpenSpp Adaptor

### PR DESCRIPTION
**Pull request by nhatnm0612**
## Summary

Add adaptor for OpenSpp

## Details

Based on npm module Odoo [https://www.npmjs.com/package/odoo].

Creating a adaptor for OpenSPP with these functions:
- [x] Write the Adaptor
- [ ] Prepare a demo flow

registrant_id is from: https://github.com/OpenSPP/openspp-registry/issues/247

Implement the methods:
- [x] getGroup(registrant_id)
- [x] getIndividual(registrant_id)
- [x] searchGroup(domain)
- [x] searchIndividual(domain)
- [x] getGroupMembers(registrant_id)
- [x] getServicePoint(name)
- [x] getPrograms(program_id) https://github.com/OpenSPP/openspp-program/issues/88
- [x] getPrograms()
- [x] getEnrolledPrograms(registrant_id)
- [x] enroll(registrant_id, program_id)
- [x] unenroll(registrant_id, program_id)
- [ ] createChangeRequest(group_id, individual_id, cr_type, data)

## Issues

[Issue](https://github.com/OpenSPP/openspp-registry/issues/248)

## Review Checklist

Before merging, the reviewer should check the following items:

- [ ] Does the PR do what it claims to do?
- [ ] If this is a new adaptor, has the migration tool been run and the
      migration guide followed?
- [ ] Are there any unit tests? Should there be?
- [ ] Is there a changeset associated with this PR? Should there be? Note that
      dev only changes don't need a changeset.

### CODE REVIEW

File: packages/openspp/src/Adaptor.js [hunk 1]
```diff
@@ -0,0 +1,261 @@
...
+ * get household information from OpenSPP
+ * @public
+ * @example
+ * getHousehold("6410000117")
+ * @function
+ * @param {string} householdName - The name of the household
+ * @param {function} callback - An optional callback function
+ * @returns {Operation}
+ */
+export function getHousehold(householdName, callback=false) {
```
Comment (jeremi) [inline packages/openspp/src/Adaptor.js]: rename to getGroup

File: packages/openspp/src/Adaptor.js [hunk 2]
```diff
@@ -0,0 +1,261 @@
...
+    login(state);
+  }
+  return state => {
+    let defaultDomain = [
+      ["is_ended", "=", false],
+      ["group.name", "=", householdName]
+    ];
+    let defaultFields = [
+      "individual", "kind", "start_date", "ended_date",
+      "individual_birthdate", "individual_gender"
```
Comment (jeremi) [inline packages/openspp/src/Adaptor.js:256]: Those field ("individual_birthdate", "individual_gender") are not in the default implementation of OpenSPP I think.
Try the code with a fresh install with openspp-docker.

File: packages/openspp/src/Adaptor.js [hunk 3]
```diff
@@ -0,0 +1,261 @@
...
+  sppConnector = new Odoo({
+    host: baseUrl,
+    database: database,
+    username: username,
+    password: password
+  });
+  sppConnector.connect((err) => {
+    if (err) {
+      Log.error(err);
+      throw new Error("Can't login to OpenSPP, please check your credentials or network!");
```
Comment (jeremi) [inline packages/openspp/src/Adaptor.js:76]: Do we have more details in the error we can give to the users?

File: packages/openspp/src/Adaptor.js [hunk 4]
```diff
@@ -0,0 +1,261 @@
...
+ * @public
+ * @example
+ * getAgentsByNumber("000117")
+ * @function
+ * @param {string} agentNumber - The number of the agent
+ * @param {number} [offset=0] - Offset searching
+ * @param {function} callback - An optional callback function
+ * @returns {Operation}
+ */
+export function getAgentsByNumber(agentNumber, offset=0, callback=false) {
```
Comment (jeremi) [inline packages/openspp/src/Adaptor.js]: It is called ServicePoint in OpenSPP.

**Review by jeremi — CHANGES_REQUESTED** [review]
- Rename Household to group everywhere.
- Add `getIndividual`
- Add `getPrograms`
- Add `getEnrolledPrograms`
- Add `enroll`
- Add `unenroll`
- Add `createChangeRequest`

File: packages/openspp/src/Adaptor.js (hunk 2 above)
Comment (nhatnm0612) [reply packages/openspp/src/Adaptor.js:256]: You are right, this is not in OpenSPP, but is implemented by you in OpenG2P repo. [(link here)](https://github.com/OpenG2P/openg2p-registry/blob/c3ff14407de55c51e64f7dc598b736a350b26e7d/g2p_registry_membership/models/group_membership.py#L31)
And I think that it is good to have.

File: packages/openspp/src/Adaptor.js (hunk 1 above)
Comment (nhatnm0612) [reply packages/openspp/src/Adaptor.js]: I changed the name

File: packages/openspp/src/Adaptor.js (hunk 4 above)
Comment (nhatnm0612) [reply packages/openspp/src/Adaptor.js]: I changed the name

**Pull request closed without merging**